private/bufpkg/bufmodule/bufmodulestore/module_data_store.go

```diff
@@ -440,13 +430,13 @@ func (p *moduleDataStore) putModuleData(
 		// Otherwise, release shared lock and upgrade to an exclusive lock for writes.
 		if readUnlocker != nil {
 			err := readUnlocker.Unlock()
-			readLocked = false
 			if err != nil {
 				return err
 			}
+			readUnlocker = nil // unset the readUnlocker since we are upgrading the lock
```

Comment:
I think we want to move this after line 432 so we don't try to unlock two times.